teamcity/tests/test_unit.py

```diff
@@ -1,47 +1,233 @@
+            id="Filter all projects",
+        ),
+        pytest.param(
+            [
+                'project1.prod',
+                'project2.prod',
+                'project1_fork.prod',
+                'project3.prod',
+                'project1.dev',
+                'project10.prod',
+                'project11_fork.prod',
+                'project18.prod',
+                'project12.prod',
+            ],
+            'projects',
+            CONFIG_ALL_BUILD_CONFIGS,
+            [],
+            [],
+            (
+                OrderedDict(
+                    [
+                        ('project1.prod', {'project1.*\\.prod': {}}),
+                        ('project1_fork.prod', {'project1.*\\.prod': {}}),
+                        ('project10.prod', {'project1.*\\.prod': {}}),
+                        ('project11_fork.prod', {'project1.*\\.prod': {}}),
+                        ('project18.prod', {'project1.*\\.prod': {}}),
+                    ]
+                ),
+                True,
+            ),
+            id="Filter all projects with limit",
+        ),
+        pytest.param(
+            ['project1.prod', 'project2.prod', 'project1.dev', 'project1_draft.dev'],
+            'projects',
+            CONFIG_ONLY_EXCLUDE_ONE_PROJECT,
+            [],
+            [],
+            (OrderedDict([('project1.prod', None), ('project2.prod', None)]), False),
+            id="Filter projects with one exclude",
+        ),
+        pytest.param(
+            ['project1.prod', 'project2.prod'],
+            'projects',
+            CONFIG_ALL_BUILD_CONFIGS_WITH_LIMIT,
+            [],
+            [],
+            (OrderedDict([('project1.prod', {"project1.*\\.prod": {"limit": 3}})]), False),
+            id="Filter projects with one include",
+        ),
+        pytest.param(
+            ['project1.prod', 'project2.prod'],
+            'projects',
+            CONFIG_MULTIPLE_PROJECTS_MAPPING,
+            [],
+            [],
+            (OrderedDict([('project1.prod', {'project1.*': {}}), ('project2.prod', {'project2.*': {}})]), False),
+            id="Filter multiple projects",
+        ),
+        pytest.param(
+            ['project1.prod', 'project2.prod', 'project2.dev', 'project1.tmp'],
+            'projects',
+            CONFIG_FILTERING_PROJECTS,
+            [],
+            [],
+            (
+                OrderedDict(
+                    [
+                        ('project1.prod', {'project1.*': {}}),
+                        ('project2.prod', {'project2.*': {}}),
+                        ('project2.dev', {'project2.*': {}}),
+                    ]
+                ),
+                False,
+            ),
+            id="Filter multiple projects with overlap and exclude override",
+        ),
+    ],
+)
+def test_filter_projects(list_to_filter, key, config, global_include, global_exclude, expected_filtered_list):
```

Comment:
```suggestion
def test_filter_projects(to_filter, key, config, global_include, global_exclude, expected_result):
    filtered = filter_items(to_filter, key, 5, global_include, global_exclude, config)
    if PY2:
        # TODO: explain why we don't check "reached limit" here.
        assert sorted(filtered[0].items()) == sorted(expected_result[0].items())
    else:
        assert filtered == expected_result
```

Motivation:
- Don't include the variable's type (`list`) in the variable name.
- Avoid reassigning a variable and changing its type (`expected_filtered_list: tuple -> expected_filtered_list: list`), this confusing to the reader.
- Decrease verbosity, that call to `filter_items` is pretty big, rather read it just once.
- Tuple unpacking for just one assertion seems like overkill.

Reply:
Slightly related, I'm wondering why `5` for the default limit in this test?

Reply:
I set the limit to `5` so that we don't have to make a large test case to validate the limit functionality